server/replicate/replicate.go

```diff
@@ -227,13 +231,35 @@ func (m *ModeManager) checkCanSync() bool {
 }
 
 func (m *ModeManager) recoverProgress() (current, total int) {
-	// FIXME: only a placeholder now. (done in 30s)
 	m.RLock()
 	defer m.RUnlock()
-	total = 300
-	current = int(time.Since(m.drAutosync.RecoverStartTime).Seconds() * 10)
-	if current > total {
-		current = total
+	var key []byte
+	for len(key) > 0 || total == 0 {
+		regions := m.cluster.ScanRegions(key, nil, 1024)
+		if len(regions) == 0 {
+			log.Warn("scan empty regions", zap.ByteString("start-key", key))
+			total++ // make sure it won't complete
+			return
+		}
+
+		total += len(regions)
+		for _, r := range regions {
+			if !bytes.Equal(key, r.GetStartKey()) {
+				log.Warn("found region gap", zap.ByteString("key", key), zap.ByteString("region-start-key", r.GetStartKey()))
+				total++
+			}
+			if r.GetReplicateStatus().GetRecoverId() == m.drAutosync.RecoverID &&
```

Comment:
StateID?

Reply:
Update it in later PR. The kvproto is not updated here.

Reply:
@disksing Do we need to check region state, invalid region should not include in `total` count?